feat(experimental): Support struct constructors in match patterns

jfecher opened the pull request:
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/7431

## Summary\*

Implements structs in match patterns

## Additional Context

There's some mostly unavoidable duplication involving checking for missing/duplicate fields when converting constructor expressions to a pattern. I think abstracting this to work with the existing already abstracted code to do so converting constructors to HirExpressions would result in less readable code.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

asterite reviewed (approved):
Looks good! It would be nice to have tests for each of the errors that happen here (duplicate field, no such field, etc.)

jfecher commented:
Added tests along with a separate quick fix that I found while writing the tests.

We were erroring with "variable already defined in pattern" when a pattern used `_` multiple times, e.g. `Foo { x: _, y: _ }`. I added a quick check to allow this and it is tested for by the new tests.
